### PR TITLE
Add support for CDI

### PR DESCRIPTION
**Pull request by almaslennikov**
https://github.com/container-orchestrated-devices/container-device-interface

### CODE REVIEW

File: images/cdi/k8s-rdma-shared-dev-plugin-cdi-ds.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      name: rdma-shared-dp-ds
+  template:
+    metadata:
+      labels:
+        name: rdma-shared-dp-ds
+    spec:
+      hostNetwork: true
+      priorityClassName: system-node-critical
+      containers:
+        - image: ghcr.io/mellanox/k8s-rdma-shared-dev-plugin:latest-amd64
```
Comment (almaslennikov) [inline images/cdi/k8s-rdma-shared-dev-plugin-cdi-ds.yaml:18]: This line is dependent on another opened PR: https://github.com/Mellanox/k8s-rdma-shared-dev-plugin/pull/75

**Comment by almaslennikov** [comment]
@abdallahyas We don't have any copyrights in the repo, should we copyright all files or drop the check?

**Comment by abdallahyas** [comment]
> @abdallahyas We don't have any copyrights in the repo, should we copyright all files or drop the check?

@almaslennikov we can not drop because it is the company policy, so we can copy the copy right message from any other project and add it to only the newly added files. The CI will only check newly added files, and would ignore the modified files

File: cmd/k8s-rdma-shared-dp/main.go [hunk 2]
```diff
@@ -28,12 +46,19 @@ func main() {
...
+	flag.BoolVar(&useCdi, "use-cdi", false,
+		"Use Container Device Interface to expose devices in containers")
 	flag.Parse()
 	if versionOpt {
 		fmt.Printf("%s\n", printVersionString())
 		return
 	}
 
+	if useCdi {
+		log.Printf("CDI enabled")
```
Comment (e0ne) [inline cmd/k8s-rdma-shared-dp/main.go:59]: log.Printf => log.Println

File: pkg/cdi/cdi.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		devices = append(devices, cdi.QualifiedName(resourceNamePrefix, resourceKind, id))
+	}
+	annoValue, err := cdi.AnnotationValue(devices)
+	if err != nil {
+		log.Printf("can not create container annotation: %q", err)
+		return nil, err
+	}
+	annotations[annoKey] = annoValue
+
+	log.Printf("Created CDI annotations: %v", annotations)
```
Comment (e0ne) [inline pkg/cdi/cdi.go:104]: plese, start log messages with lower-case letter

**Review by e0ne — CHANGES_REQUESTED** [review]

**Comment by almaslennikov** [comment]
@e0ne addressed your comments and rebased the PR, PTAL

**Review by e0ne — APPROVED** [review]
LGTM. thanks for addressing my comments

File: pkg/cdi/cdi.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		log.Printf("createCDISpec(): can not create CDI json: %q", err)
+		return err
+	}
+
+	return nil
+}
+
+// CreateContainerAnnotations creates container annotations based on CDI spec for a container runtime
+func (c *Impl) CreateContainerAnnotations(
+	devicesIDs []string, resourceNamePrefix, resourceKind string) (map[string]string, error) {
```
Comment (adrianchiris) [inline pkg/cdi/cdi.go:86]: fail if deviceIDs is empty ?

File: pkg/cdi/cdi_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+	"github.com/Mellanox/k8s-rdma-shared-dev-plugin/pkg/cdi"
+)
+
+func TestCdi(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "CDI Suite")
+}
+
+var _ = Describe("Cdi", func() {
```
Comment (adrianchiris) [inline pkg/cdi/cdi_test.go:35]: can you test CreateCDISpec() ?

File: pkg/cdi/cdi_test.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+			deviceID := "0000:00:00.1"
+			cdi := cdi.Impl{}
+			annotations, err := cdi.CreateContainerAnnotations([]string{deviceID}, "example.com", "net")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(annotations)).To(Equal(1))
+			annoKey := "cdi.k8s.io/example.com_net"
+			annoVal := "example.com/net=0000:00:00.1"
+			Expect(annotations[annoKey]).To(Equal(annoVal))
+		})
+		It("should not fail on non-existing device", func() {
```
Comment (adrianchiris) [inline pkg/cdi/cdi_test.go:47]: what is the difference between this test case and the first one ?

File: pkg/resources/server.go [hunk 7]
```diff
@@ -16,20 +34,23 @@ import (
...
 	"github.com/Mellanox/k8s-rdma-shared-dev-plugin/pkg/types"
 )
 
 const (
 	// Local use
-	cDialTimeout  = 5 * time.Second
-	watchWaitTime = 5 * time.Second
+	cDialTimeout    = 5 * time.Second
+	watchWaitTime   = 5 * time.Second
+	cdiResourceKind = "net-rdma"
```
Comment (adrianchiris) [inline pkg/resources/server.go:45]: nit: net-shared-rdma ?

File: pkg/resources/server.go [hunk 8]
```diff
@@ -275,6 +301,12 @@ func (rs *resourceServer) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlu
 		return err
 	}
 
+	err := rs.updateCDISpec()
```
Comment (adrianchiris) [inline pkg/resources/server.go:304]: why do you prefer to create the spec here an not on allocate when required ?

File: pkg/resources/server.go [hunk 9]
```diff
@@ -291,10 +323,28 @@ func (rs *resourceServer) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlu
 				rs.updateResource <- true
 				return err
 			}
+			err := rs.updateCDISpec()
```
Comment (adrianchiris) [inline pkg/resources/server.go:326]: if you do it here, you need to protect using rs.mutex like done in rs.sendDevices()

File: pkg/resources/server.go [hunk 10]
```diff
@@ -291,10 +323,28 @@ func (rs *resourceServer) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlu
...
 }
 
+func (rs *resourceServer) updateCDISpec() error {
+	// check if CDI mode is enabled
+	if !rs.useCdi {
+		return nil
+	}
+	err := rs.cdi.CreateCDISpec(rs.resourcePrefix, cdiResourceKind, rs.devs, rs.deviceSpec)
+	if err != nil {
+		log.Printf("initServers(): error creating CDI spec: %v", err)
```
Comment (adrianchiris) [inline pkg/resources/server.go:342]: this is called within ListAndWatch ... log msg is missleading

File: pkg/resources/server_test.go [hunk 11]
```diff
@@ -57,7 +75,7 @@ var _ = Describe("resourceServer tests", func() {
 			}
 			defer fs.Use()()
 			conf := &types.UserConfig{ResourceName: "test_server", ResourcePrefix: "rdma", RdmaHcaMax: 100}
-			obj, err := newResourceServer(conf, fakeDeviceList, true, "socket")
+			obj, err := newResourceServer(conf, fakeDeviceList, true, "socket", false)
```
Comment (adrianchiris) [inline pkg/resources/server_test.go:78]: lets add tests for:

1. ListAndWatch changes
2. Allocate changes

**Review by adrianchiris — COMMENTED** [review]
Hey @almaslennikov appologies for the delayed review !

**Comment by adrianchiris** [comment]
i have deployed the changes, something is off with the cdi spec, expect further feedback :)

File: pkg/resources/server.go (hunk 8 above)
Comment (adrianchiris) [reply pkg/resources/server.go:304]: thinking on it, i think its fine to keep it here. that is, CDI spec generation is performed in listAndWatch

**Comment by adrianchiris** [comment]
I have deployed device plugin in cdi mode.

the cdi specification generated is as follows:

```
cdiVersion: 0.5.0
containerEdits: {}
devices:
- containerEdits:
    deviceNodes:
    - hostPath: /dev/infiniband/issm0
      path: /dev/infiniband/issm0
      permissions: rw
    - hostPath: /dev/infiniband/umad0
      path: /dev/infiniband/umad0
      permissions: rw
    - hostPath: /dev/infiniband/uverbs0
      path: /dev/infiniband/uverbs0
      permissions: rw
    - hostPath: /dev/infiniband/rdma_cm
      path: /dev/infiniband/rdma_cm
      permissions: rw
    - hostPath: /dev/infiniband/issm1
      path: /dev/infiniband/issm1
      permissions: rw
    - hostPath: /dev/infiniband/umad1
      path: /dev/infiniband/umad1
      permissions: rw
    - hostPath: /dev/infiniband/uverbs1
      path: /dev/infiniband/uverbs1
      permissions: rw
    - hostPath: /dev/infiniband/rdma_cm
      path: /dev/infiniband/rdma_cm
      permissions: rw
  name: "0"
- containerEdits:
    deviceNodes:
    - hostPath: /dev/infiniband/issm0
      path: /dev/infiniband/issm0
      permissions: rw
    - hostPath: /dev/infiniband/umad0
      path: /dev/infiniband/umad0
      permissions: rw
    - hostPath: /dev/infiniband/uverbs0
      path: /dev/infiniband/uverbs0
      permissions: rw
    - hostPath: /dev/infiniband/rdma_cm
      path: /dev/infiniband/rdma_cm
      permissions: rw
    - hostPath: /dev/infiniband/issm1
      path: /dev/infiniband/issm1
      permissions: rw
    - hostPath: /dev/infiniband/umad1
      path: /dev/infiniband/umad1
      permissions: rw
    - hostPath: /dev/infiniband/uverbs1
      path: /dev/infiniband/uverbs1
      permissions: rw
    - hostPath: /dev/infiniband/rdma_cm
      path: /dev/infiniband/rdma_cm
      permissions: rw
  name: "1"
kind: rdma/net-rdma
```
the list of devices goes on up until `rdmaHcaMax` num is reached.

A couple of comments on this one:

1. the device kind is composed of vendor prefix and device class. in our case we should use `nvidia.com` as vendor prefix.
for now as hardcoded value, and later on it may be configurable if desired. current vendor prefix, `rdma` which is what we have hardcoded as resource prefix in device plugin does not really fit in CDI kind. device class `net-rdma` fits well.

2. devices here are eventually PCI devices. so i think CDI device spec should reflect the actual device, its name should be the PCI device address. having duplicate devices (notice also there is duplicate mounts per device) here does not make much sense since eventually we can provide the same CDI device name in multiple device plugin Allocate calls. that means, we would a CDI device per PCI device selected. on Allocate call we may return either one or multiple annotations pointing to one or several CDI devices depending on the actual devices that are part of the pool.

4. current code does not support more than one resource pool since you are writing to the same file `/var/run/cdi/rdma.yaml`


example of a cdi spec file as i envision it:

```
cdiVersion: 0.5.0
containerEdits: {}
devices:
- containerEdits:
    deviceNodes:
    - hostPath: /dev/infiniband/issm0
      path: /dev/infiniband/issm0
      permissions: rw
    - hostPath: /dev/infiniband/umad0
      path: /dev/infiniband/umad0
      permissions: rw
    - hostPath: /dev/infiniband/uverbs0
      path: /dev/infiniband/uverbs0
      permissions: rw
    - hostPath: /dev/infiniband/rdma_cm
      path: /dev/infiniband/rdma_cm
      permissions: rw
  name: "0000:03:00.0"
- containerEdits:
    deviceNodes:
    - hostPath: /dev/infiniband/issm1
      path: /dev/infiniband/issm1
      permissions: rw
    - hostPath: /dev/infiniband/umad1
      path: /dev/infiniband/umad1
      permissions: rw
    - hostPath: /dev/infiniband/uverbs1
      path: /dev/infiniband/uverbs1
      permissions: rw
    - hostPath: /dev/infiniband/rdma_cm
      path: /dev/infiniband/rdma_cm
      permissions: rw
  name: "0000:03:00.1"
kind: nvidia.com/net-rdma
```

File: pkg/resources/server.go [hunk 12]
```diff
@@ -16,20 +34,23 @@ import (
...
+	watchWaitTime   = 5 * time.Second
+	cdiResourceKind = "net-rdma"
 )
 
 type resourcesServerPort struct {
 	server *grpc.Server
 }
 
 type resourceServer struct {
+	resourcePrefix string
```
Comment (adrianchiris) [inline pkg/resources/server.go:53]: you may have the same resource prefix for multiple resource servers

**Review by adrianchiris — CHANGES_REQUESTED** [review]

File: pkg/resources/server.go [hunk 13]
```diff
@@ -291,10 +326,28 @@ func (rs *resourceServer) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlu
...
+			err := rs.updateCDISpec()
+			if err != nil {
+				log.Printf("cannot update CDI specs: %v", err)
+				return err
+			}
 		}
 	}
 }
 
+func (rs *resourceServer) updateCDISpec() error {
```
[the remark this answers is not in the document]
Comment (e0ne) [reply pkg/resources/server.go:338]: looks like so. we need to find a way to resolve this issue.

Maybe we can create spec files with a prefix (e.g. `rdma-shared-dp-<pool-name>.yaml`) and delete all these files in the begining of `updateCDISpec`

[the remark this answers is not in the document]
Comment (adrianchiris) [reply pkg/resources/server.go:338]: also there is the question of what happens if we change the devices of an existing resource pool, will it replace or just append, i didnt check this.

[the remark this answers is not in the document]
Comment (almaslennikov) [reply pkg/resources/server.go:340]: When we change the resource pool, the new spec overwrites the previous one, so we don't need to handle this case explicitly. I added a cleanup call when starting the resource servers to account for the case when the previous instance of device plugin finished unsuccessfully

File: pkg/cdi/cdi.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+// CDI represents CDI API required by Device plugin
+type CDI interface {
+	CreateCDISpec(resourcePrefix, resourceName, poolName string, devices []types.PciNetDevice) error
+	CreateContainerAnnotations(
+		devices []types.PciNetDevice, resourcePrefix, resourceKind string) (map[string]string, error)
+}
+
+// Impl implements CDI interface
+type Impl struct {
```
Comment (e0ne) [inline pkg/cdi/cdi.go:41]: let's make this structure private

**Review by e0ne — CHANGES_REQUESTED** [review]

File: deployment/k8s/overlay/cdi.yaml [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+metadata:
+  name: rdma-shared-dp-ds
+  namespace: kube-system
+spec:
+  template:
+    spec:
+      containers:
+        - name: k8s-rdma-shared-dp-ds
+          volumeMounts:
+            - name: default-cdi
```
Comment (e0ne) [inline deployment/k8s/overlay/cdi.yaml:12]: no need to mount `/etc/cdi`

**Review by e0ne — CHANGES_REQUESTED** [review]

**Review by e0ne — APPROVED** [review]

**Pull request merged**